onnxruntime/core/framework/inline_containers.h

```diff
@@ -0,0 +1,192 @@
+  const SafeInt<size_t> slot_offset = (num_control_bytes + slot_size - 1) & (~slot_size + 1);
+  return (slot_offset + nelem * slot_size);
+}
+
+}  // namespace inline_containers_internal
+
+/// <summary>
+/// Estimate memory requirements for an InlinedHashSet
+/// so it can be pre-allocated on a stack or using other allocator when the number
+/// of elements is known. This provides an oppty to bring the number of allocations
+/// down to zero.
+/// The InlinedHashSet keeps values in the buckets array which is allocated in one shot.
+/// </summary>
+/// <param name="num_elements">number of elements</param>
+/// <returns></returns>
+template <class T>
+inline size_t EstimateInlinedHashSetMemory(size_t num_elements) {
+  constexpr size_t slot_size = sizeof(InlinedHashSet<T>::slot_type);
+  return inline_containers_internal::EstimateHashStorageSize(slot_size, num_elements);
+}
+
+/// <summary>
+/// Estimate memory requirements for an InlinedHashMap
+/// so it can be pre-allocated on a stack or using other allocator when the number
+/// of elements is known. This provides an oppty to bring the number of allocations
+/// down to zero.
+/// The InlinedHashMap keeps values in the buckets array which is allocated in one shot.
+/// </summary>
+/// <param name="num_elements">number of elements</param>
+/// <returns></returns>
+template <class K, class V>
+inline size_t EstimateInlinedHashMapMemory(size_t num_elements) {
+  constexpr size_t slot_size = sizeof(InlinedHashMap<K, V>::slot_type);
+  return inline_containers_internal::EstimateHashStorageSize(slot_size, num_elements);
+}
+
+inline bool IsSizeOverStackAllocationLimit(size_t size) {
+  return size > kOrtStackAllocationLimitBytes;
+}
+
+// This macro can serve to pre-allocate estimated memory size for a container.
+// The allocation would take place on a stack up to a certain limit. With proper calculation
+// one can utilize the stack for small maps/sets for both pmr::Inlined and std::pmr::map/unordered_map
+// std::pmr::set/onordered_set. For std containers, this can reduce to 1 or even may completely eliminate
+// the number of new/delete calls.
+#define OrtDeclareAllignedStackOrAllocatedBuffer(buffer_ptr, size_in_bytes)                          \
+  std::unique_ptr<uint8_t[]> on_heap_##buffer_ptr;                                                   \
+  void* buffer_ptr = (size_in_bytes > kOrtStackAllocationLimitBytes)                                 \
+                         ? inline_containers_internal::allocate(size_in_bytes, on_heap_##buffer_ptr) \
+                         : ORT_ALLOCA(size_in_bytes)
+
+// This gives a set size stackbuffer
+template <typename T, size_t N>
+class SmallBuffer {
+  T buffer_[N];
+
+ public:
+  T* Buffer() noexcept { return buffer_; }
+  constexpr size_t size() const noexcept { return N; }
+  constexpr size_t size_in_bytes() const noexcept { return sizeof(T) * N; }
+};
+
+class SmallBufferResource {
+  std::pmr::monotonic_buffer_resource resource_;
```

Comment:
Any reason to not derive SmallBufferResource from monotonic_buffer_resource (similar to DebugMemoryResource)?